backend/src/main/java/fk/prof/backend/model/policy/ZookeeperBasedPolicyStoreAPI.java

```
@@ -0,0 +1,232 @@
+        ZookeeperUtil.sync(curatorClient, policyPath).setHandler(ar -> {
+            if (ar.failed()) {
+                syncException.initCause(ar.cause());
+            }
+            syncLatch.countDown();
+        });
+
+        boolean syncCompleted = syncLatch.await(10, TimeUnit.SECONDS);
+        if (!syncCompleted) {
+            throw new PolicyException("ZK sync operation taking longer than expected", true);
+        }
+        if (syncException.getCause() != null) {
+            throw new PolicyException(syncException, true);
+        }
+        for (String appId : curatorClient.getChildren().forPath(policyPath)) {
+            for (String clusterId : curatorClient.getChildren().forPath(policyPath + DELIMITER + appId)) {
+                for (String procName : curatorClient.getChildren().forPath(policyPath + DELIMITER + appId + DELIMITER + clusterId)) {
+                    String zNodePath = policyPath + DELIMITER + appId + DELIMITER + clusterId + DELIMITER + procName;
+                    PolicyDTO.PolicyDetails policyDetails = PolicyDTO.PolicyDetails.parseFrom(ZookeeperUtil.readLatestSeqZNodeChild(curatorClient, zNodePath));
+                    inMemoryPolicyCache.put(decode32(appId), decode32(clusterId), decode32(procName), policyDetails);
+                }
+            }
+        }
+    }
+
+    void init() throws Exception {
+        synchronized (this) {
+            if (!initialized) {
+                populateCacheFromZK();
+                initialized = true;
+            }
+        }
+    }
+
+    Set<String> getAppIds(String prefix) {
+        if (prefix == null) return new HashSet<>();
+        Set<String> appIds = inMemoryPolicyCache.getAppId();
+        return appIds.stream().filter(s -> s.startsWith(prefix)).collect(Collectors.toSet());
+
+    }
+
+    Set<String> getClusterIds(String appId, String prefix) {
+        if (prefix == null) return new HashSet<>();
```

what if app id is null

---

Thought about it, but skipped it assuming it will be implementable as fail fast checks at API level. But will consider full proofing this layer as well for null checks.

---

assert NPE is thrown, basically checks at some layer to handle null params. If yes, remove null checks here